src/attributes/classes.js

```diff
@@ -100,60 +99,26 @@ jQuery.fn.extend( {
 	},
 
 	toggleClass: function( value, stateVal ) {
-		var type = typeof value;
+		var type = typeof value,
+			classNames = type === "string" ? value.match( rnotwhite ) : [];
 
-		if ( typeof stateVal === "boolean" && type === "string" ) {
-			return stateVal ? this.addClass( value ) : this.removeClass( value );
-		}
-
-		if ( jQuery.isFunction( value ) ) {
-			return this.each( function( i ) {
-				jQuery( this ).toggleClass(
-					value.call( this, i, getClass( this ), stateVal ),
-					stateVal
-				);
-			} );
-		}
-
-		return this.each( function() {
-			var className, i, self, classNames;
-
-			if ( type === "string" ) {
+		return this.each( function( i ) {
+			var className,
+				self = jQuery( this ),
+				c = 0;
 
-				// Toggle individual class names
-				i = 0;
-				self = jQuery( this );
-				classNames = value.match( rnotwhite ) || [];
-
-				while ( ( className = classNames[ i++ ] ) ) {
-
-					// Check each className given, space separated list
-					if ( self.hasClass( className ) ) {
-						self.removeClass( className );
-					} else {
-						self.addClass( className );
-					}
-				}
-
-			// Toggle whole class name
-			} else if ( value === undefined || type === "boolean" ) {
-				className = getClass( this );
-				if ( className ) {
+			if ( type === "function" ) {
+				classNames = value.call( this, i, getClass( this ), stateVal )
+					.match( rnotwhite ) || [];
+			}
 
-					// Store className if set
-					dataPriv.set( this, "__className__", className );
-				}
+			// Toggle individual class names based on presence or stateVal
+			while ( ( className = classNames[ c++ ] ) ) {
 
-				// If the element has a class name or if we're passed `false`,
-				// then remove the whole classname (if there was one, the above saved it).
-				// Otherwise bring back whatever was previously saved (if anything),
-				// falling back to the empty string if nothing was stored.
-				if ( this.setAttribute ) {
-					this.setAttribute( "class",
-						className || value === false ?
-						"" :
-						dataPriv.get( this, "__className__" ) || ""
-					);
+				if ( stateVal === false || stateVal !== true && self.hasClass( className ) ) {
```

Comment:
In `.toggleClass()` the `stateVal` can be either `true`, `false` or `undefined`. The boolean cases indicate the caller wants to force the class to be added or removed. For the `undefined` case, the class should be added if it does not have the class and removed if it does have the class. So if you walk through the three states you can double-check that expression works for all the cases. I added a few extra tests below to catch some of the possible mistakes someone could make in editing this line. :smile_cat: